### ISSUE TITLE
Running simulations in parallel

### ISSUE THREAD

**Issue by SachinBM-CE** [NVIDIA/TorchFort#60]
Hi !
Is it possible to run multiple simulations in parallel, all updating the same policy / critic network using TorchFort?
If so can you please share some insights into this?
Currently I'm launching one simulation using one set of - (binary, case file)
What if there are multiple of these (binary, case file), but having the same yaml file and torchscript?

**Comment by SachinBM-CE** [comment]
I will wait for a response from @azrael417 regarding the multi-env / multi-agent.

1. By running multiple independent models concurrently, do you mean each MPI process is having it's own copy of the neural network? 
2. Is `update_replay_buffer` also affected by `create_distributed_system` ? Is the replay buffer having it's local replay buffer corresponding to each MPI Process?
3. How does the `train_step` vary with `create_distributed_system` and `create_system`?

The `update_replay_buffer` is checking if the `batch_dimension` is equal or not to `n_envs` parameter. The `batch_dimension` which is also equal to the `number of agents` may vary depending on the number of MPI processes and different zones of the domain. But what remains fixed is the total number of agents across all processes or zones. 

**For example:** Suppose if 2 MPI processes are launched, following data is observed
```
----------------------------------------
Rank, boundary zone, number of agents
----------------------------------------
   0,             1,               m
   0,             2,               n
   1,             1,               m
   1,             2,               n
```
Should I do a `MPI_Gather` operation, to collect (s,a,r,s') from all the processes/zones whose sum total is always fixed, which can be used directly to set the `n_envs` parameter in the `config.yaml`? (OR) is there any functionality in TorchFort that can help in this scenario?

P.S.: Just for context, the dimension of s, a, r, s' are respectively - (number of agents, number of features), (number of agents, 1), (number of agents) and (number of agents, number of features). Here the number of agents can be equal to **m** or **n** depending on the boundary zone.

**Comment by azrael417** [comment]
Hello Sachin,

multi does not refer to multi agent but instead multi environment. It allows you to run multiple environments in a single task, but those apply to the same agents. In order to achieve multi agents, you should be able to take care of the independent agents in the backend. What is the actual setup you are using? Do you for example have two boundary layers and want to apply a different agent to each boundary layer? What is the actual model for each agent? An MLP acting on a feature vector and you want to have a different MLP for each wallpoint?

Concerning the gather you do not need to do that, if you initialized the distributed system accordingly all the gradient communication should be taken care of automatically. But before we go there, we need to understand what you are trying to achieve. For example for each agent, you might to create its own distributed system since inside a distributed system, all gradients are shared (which is what you want to avoid when running independent agents). 

Best
Thorsten

**Comment by SachinBM-CE** [comment]
Hi Thorsten,
The MPI_Gatherv implementation on rank 0 worked ! By collecting global (s,a,r,s') arrays from all ranks, I've eliminated the uneven agent distribution problem entirely. The replay buffer updates now handle the complete dataset regardless of how agents are spread across processes. Thanks for the guidance on this solution!

**Comment by SachinBM-CE** [comment]
I am collecting 100 timesteps of simulation data (s_1, a_1 , r_1, s'_1) ... (s_100, a_100, r_100, s'_100) which is equivalent to one episode (as terminal state is reached) and after which I want to train the MLP. 

How do I set the `epochs` and `number of gradient updates` for training? Currently I have placed the call to train_step function inside a do loop. How is it ensured that entire episode data is trained?

**Comment by SachinBM-CE** [comment]
sure, I'm using off-policy Soft-Actor-Critic for my training.
I'm having some problems relating to normalization of the state and reward... 
We had previously discussed about this (https://github.com/NVIDIA/TorchFort/issues/48#issuecomment-2969989589)...
Suppose my states are (u^+, du^+/dy^+, y^+) all in wall units and I have some relative error based reward function, do I have to take the running mean and standard deviation of these quantities over certain time frame? How is it usually done?

**Comment by azrael417** [comment]
Hello Sachin, this is part of the research I guess. However, I would refrain from using changing normalizations because especially with off policy algorithms, you want to be able to compare old state transitions to newer ones and if you change normalizations in between (which you would do if you use running stats) then the comparison is not meaningful.

In that sense, I would thermalize the system without actuation (basically get in in a steady state), compute all stats there (for mean/var of the fields and mean of the reward), then use that throughput the entire training.

If you would do some fine-tuning later, you could think about recomputing those on the actuated system, but I would start with the former.

**Comment by SachinBM-CE** [comment]
Hi ! 
I was trying to run my application on HPC, but I get the following segmentation fault -
```
Thread 1 "neko" received signal SIGSEGV, Segmentation fault.
0x00001555430cfb33 in mkl_blas_avx512_dgemv_t_intrinsics () from /path/to/env/lib/python3.12/site-packages/torch/lib/libtorch_cpu.so
Missing separate debuginfos, use: yum debuginfo-install glibc-2.28-251.el8_10.25.x86_64
```
I did not face this issue on my local machine, but it occurs on the HPC. 
I loaded the following modules:
`module load python/3.12-conda gcc/11.2.0 openmpi/4.1.2-gcc11.2.0 cmake/3.24.3 hdf5/1.14.5-ompi-gcc mkl/2021.4.0`

**Comment by SachinBM-CE** [comment]
Can you please look into the last error (posted above) when working with PPO. Thank you.

**Comment by azrael417** [comment]
Concerning PPO, the line you are pointing at is just a bias parameter for the previous layer. I think there is a typo there

instead of 

```actor_biases.push_back(register_parameter("actor_b_entry", torch::zeros(encoder_layer_sizes[0])));```

can you try

```actor_biases.push_back(register_parameter("actor_b_entry", torch::zeros(actor_layer_sizes[0])));```

and let me know if that runs?

**Comment by azrael417** [comment]
As you can see from below, the second parameter in linear is the number of output features and the bias has to have the same shape:

```for (int i = 0; i < actor_layer_sizes.size() - 2; ++i) {
    actor_layers.push_back(register_module("actor_fc_" + std::to_string(i),
                                           torch::nn::Linear(actor_layer_sizes[i], actor_layer_sizes[i + 1])));
    actor_biases.push_back(register_parameter("actor_b_" + std::to_string(i), torch::zeros(actor_layer_sizes[i + 1])));
  }```

here it looks correct.

**Comment by SachinBM-CE** [comment]
Thank you ! I changed the code as per your suggestion, now it works fine.

The predict function throws the following error now:
```
terminate called after throwing an instance of 'c10::Error'
  what():  mat1 and mat2 shapes cannot be multiplied (12800x2 and 8x1)
```

I am using the following configuration for my multi-agent RL case:
```
rollout_buffer:
  type: gae_lambda
  parameters:
    size: 12800
    n_envs: 12800

actor_critic_model:
  type: ActorCriticMLP
  parameters:
    dropout: 0.0
    encoder_layer_sizes: [2, 16, 2] 
    actor_layer_sizes: [2, 8, 1] 
    value_layer_sizes: [2, 8, 1] 
    state_dependent_sigma: False
    log_sigma_init: 0.
```

The idea behind the above config was that from the config templates, I assumed that the output of the encoder branches into the actor network and critic network. Also, does PPO support the multi-agent case?

**Comment by SachinBM-CE** [comment]
Can you please check if any of these is causing the above issue:

https://github.com/NVIDIA/TorchFort/blob/b6353c4594a1ca4816ea94ee55ef66e391bf0be3/src/csrc/models/rl/actor_critic_model.cpp#L113

https://github.com/NVIDIA/TorchFort/blob/b6353c4594a1ca4816ea94ee55ef66e391bf0be3/src/csrc/models/rl/actor_critic_model.cpp#L105

**Comment by azrael417** [comment]
Also, what are the other settings, are you using squashing or rescaling? And can you please tell me where it fails, I need to see the full error message.

**Comment by SachinBM-CE** [comment]
Thank you very much! I will check it out and let you know!

**Comment by SachinBM-CE** [comment]
I tested this MR on my case and it works now without any shape errors. 🚀

Regarding PPO, it runs smoothly until 7 ranks, but gets killed (signal 9) when I use 8 ranks. I used the following rollout_buffer config:
```
rollout_buffer:
  type: gae_lambda
  parameters:
    size: 12800
    n_envs: 12800
```

Is there a limitation on the `size` parameter in the rollout buffer? If I increase the trajectory length to 10 resulting in size=12800x10, the program gets killed irrespective of the number of ranks used.

**Comment by azrael417** [comment]
It creates one RB per rank which means when the size gets too big and it does not fit into your CPU memory anymore the process will get killed. Are you running on multiple ranks per node?

**Comment by SachinBM-CE** [comment]
This was tested on my local machine, but I haven't run it on the cluster yet. On the cluster, I'll definitely be running with multiple ranks per node.
For context: based on our previous discussions, I'm gathering all arrays using MPI_Gatherv, sending them to TorchFort via update_rollout_buffer, and then scattering the actions back using MPI_Scatterv. This means the simulation runs across multiple ranks, but training only occurs on the root rank. This approach worked well for SAC with a much larger replay buffer (12800 × 100), but I'm now encountering this issue with PPO.
On a positive note, PPO training (nstep=1) appears to be significantly faster than SAC training (nstep=1), which is encouraging.

**Comment by SachinBM-CE** [comment]
Hi Thorsten !
Process killed behavior is not observed when the RL system (using `torchfort_rl_on_policy_create_system`) is created only on the root rank. I had missed this somehow. Now the simulation runs on multiple ranks as before and the training on the root rank.

**Comment by SachinBM-CE** [comment]
- I don't see much speedup in training when using `model_device=0` and `rb_device=0` instead of the CPU (-1). How do I check whether the train_step is taking place on GPU or not? 
- What should I assign the `stream` argument to?

**Comment by SachinBM-CE** [comment]
The above results were without saving the checkpoint. 
In the meantime, I commented the wandb_log. This reduced the time by about 30 msec. 

I will properly check again if there are some other parts drawing the compute time.

**Comment by azrael417** [comment]
In the code above, the timing included the checkpoint saving. In any case, how much does that training take compared to the simulation which is running next to it?

**Comment by SachinBM-CE** [comment]
Simulation (i.e. solving momentum and pressure equations) account for about 50% and the reinforcement learning wall model is accounting for the other 50% considering training step with just one iteration. 
I observe that time for a single step of simulation (solver + 1 train iter of RL) took 210 msec. Within this RL took about 110 msec and the solver took 100 msec. 
But if we increase the training iterations, batch_size and nstep reward reduction it will also increases the time taken for training, right?

**Comment by azrael417** [comment]
Hello Sachin,

I see the issue. Are you adding new states after each solver step and also training then? This is usually not a good idea, since you will be changing actuations too quickly. It is better to apply an actuation and then let it act for some number of solver steps. Ideally, you define a physical time interval dt after which you update the actuation. The state s and s' should also be chosen as such. If the simulation uses dynamical time, make sure to change that logic that it a least makes dt sized time steps. How long those should be is system dependent. But this is likely the reason why the RL doesn't work as you reported earlier. 

You need to:
1) thermalize the unactuated system for some physical time t_warm before you compute stats 
2) then run the unactuated system for some other time t_stats over which you measure average reward and input normalization stats
These two steps you can skip when you load a thermalized config and stats from, disk.
3) then, save state s, apply actuation a run for dt steps and save s' and r. Make dt long enough so that the actuation can actually kick in and influence the system sufficiently, but not too large because otherwise it is not adaptive enough.
Rinse and repeat.
4) fill the replay buffer with (s,a,s',r,d) tuples to a min size. Not sure what is good here, you need to experiment.
5) start training then but I do not train after every dt steps, I train after 10 dt or 100dt steps (the RB should have 10-100 new samples by then).

This all requires a lot of tuning and testing what works.

Best
Thorsten

**Comment by SachinBM-CE** [comment]
Hi Thorsten,
I was able to run inference using `predict` after loading the checkpoint from training, and switching to single precision gave satisfactory results. However, running inference on higher Reynolds numbers with finer meshes fails due to a mismatch in n_envs between the replay buffer and the checkpoint, causing the following error:
```
terminate called after throwing an instance of 'std::runtime_error'
what(): UniformReplayBuffer::update: the size of the leading dimension of tensors s, a and sp has to be equal to the number of environments
```
Would it be possible to include the graph information in the `model.pt` checkpoint so we can reconstruct the mean and std heads of the policy network and perform inference using Gaussian sampling and action clipping. So that `inference` method can be called independent of the replay buffer?
